### ISSUE TITLE
Update entity with a nullable relationship

### ISSUE THREAD

**Issue by YaLehin** [json-api-dotnet/JsonApiDotNetCore#1002]
My entity has a relationship to another entity, that allows null. And it is null by default. Then if I try to send a PATCH request to update this relationship on an existing item, I get a response with an error, that the primary entity does not exist. It seems that a problem is that when querying the primary resource for an update in queryLayer included that relationship that at the current time is null, so the result of the query is also null.
Version JADNC is 4.1.1
File: JsonApiResourceService
procedure: private async Task<TResource> GetPrimaryResourceForUpdateAsync
_repositoryAccessor.GetForUpdateAsync returns null, so AssertPrimaryResourceExists(resource) fails.
It seems to be a bug, but I may be wrong. Please in advise.

**Comment by bkoelman** [comment]
I suspect you're trying to update the wrong relationship. In JADNC, relationships have a direction. For example, the relationship Author from Book to Person is another relationship than the one from Person to Books. In a Patch Relationship request, the URL parameters identify which relationship to update (the left side, which is the primary resource type and id) and the body contains what to assign to it (the right side).

A Patch Relationship request produces a QueyLayer with an include for the existing right side. If that include is null, the left side (primary resource) is still retrieved. Unless it does not exist, which makes sense because you cannot change the right side of nothing.

**Comment by YaLehin** [comment]
Thank you for your clarification.
But it seems something wrong with that include. If include does not exists, the primary resource returned as null. But if include exists, the same PATCH query works fine and correct and the relationship is updated. May some other parameters influence the result of QueryLayer that I missed?

**Comment by bkoelman** [comment]
Can you post your resource classes and EF Core mappings?

**Comment by YaLehin** [comment]
```
public class Tag : Identifiable
{
    [Attr(PublicName = "name")]
    public string Name { get; set; } = "";
    [Attr(PublicName = "descr")]
    public string Descr { get; set; } = "";
    public int ProjectId { get; set; }
    [HasOne(PublicName="project")]
    public Project Project { get; set; }
    [HasMany(PublicName = "identities")]
    public List<Identity> Identities { get; set; } = new List<Identity>();
    public int ActuatorId { get; set; }
    [HasOne(PublicName="actuator")]
    public Actuator? Actuator { get; set; }
    [NotMapped] public int Key { get; set; } = 0;
}
```
```
public class Actuator : Identifiable
{
    [Attr(PublicName="name")]
    public string Name { get; set; }
    [Attr(PublicName="type")]
    public string Type { get; set; }
    public int ProjectId { get; set; }
    [HasOne(PublicName="project")]
    public Project Project { get; set; }
    [HasMany(PublicName = "actuatorparams")]
    public List<Actuatorparams> Actuatorparams { get; set; } = new List<Actuatorparams>();
    [HasMany(PublicName = "tags")]
    public List<Tag> Tags { get; set; } = new List<Tag>();
}
```
So, Tag has a nullable relationship to Actuator (HasOne), and Actuator has HasMany relationship to Tag.
I try to update a relationship on Tag, so the url for **PATCH** is 
http://localhost:5000/api/v1/tags/33
The body of the request is
```
{
    "data": {
        "id": "33",
        "attributes": {
            "name": "EUV201",
            "descr": "Sream to  H-201"
        },
        "relationships": {
            "actuator": {
                "data": {
                    "type": "actuators",
                    "id": "1"
                }
            }
        },
        "type": "tags"
    }
}
```
On this request (if the current state of Tag.actuator is null), I receive the error message:
```
{
    "errors": [
        {
            "id": "3d37ed37-39fc-4e13-9973-7d8d858a9c76",
            "status": "404",
            "title": "The requested resource does not exist.",
            "detail": "Resource of type 'tags' with ID '33' does not exist."
        }
    ]
}
```

**Comment by bkoelman** [comment]
Thanks. I'm not seeing anything unusual.

Which version of .NET and EF Core are you using, and what database?

And can you please paste your DbContext class and the produced SQL?

To make SQL visible in logs, add this to appsettings.json:
```json
{
  "Logging": {
    "LogLevel": {
      "Microsoft.EntityFrameworkCore.Database.Command": "Information"
    }
  }
}
```

**Comment by YaLehin** [comment]
I am using .NET and EntityFramework version 5
During development database is InMemory.
And it seems that problem may be in the database provider because logging is not working out of the box.
I'll try to investigate. Thanks.

**Comment by bkoelman** [comment]
For InMemory there is no translation to SQL because there is no database engine running. The InMemory provider is very limited in what it supports. Can you try Sqlite instead? It is easy to setup, see our GettingStarted project.

**Comment by YaLehin** [comment]
The problem was in InMemory provider. After changing to Sqlite, the error is disappeared.
Thank you very much for your help!
I think this issue may be closed.

**Comment by bart-degreed** [comment]
I've tried to repro using InMemory provider and found the root cause. Using your models, when saving a Tag to the in-memory database, it does not actually get saved. But to error is produced either (I don't know why). So when trying to update later, the Tag cannot be found. The reason why it is not saved, is because your EF Core relationship is configured incorrectly:

```c#
public int ActuatorId { get; set; }
public Actuator? Actuator { get; set; }
```

This declares `Actuator` as nullable, yet its foreign key `int ActuatorId` indicates it is required. Changing it to `int?` (or removing the foreign key) makes it work as expected.

**Comment by YaLehin** [comment]
Thank you for your investigation. I also found the error with incorrect Keys during switching to SQLite. But don't found out that it was the root cause.
I also don't have any errors from the InMemory provider, that strange. And if I send a GET request like _tags/33_ I get the correct response.
The one more strange thing was the request _tags/33/actuator_, which returns 404 error. I found the closed issue with this topic and it seems that the planned answer must be null for the actuator. And now based on your explanation I understand why was that behavior.
Thanks a lot to you, my understanding of this topic is deeply increased.
